File: 0238-product-of-array-except-self/python/Solution.py
class Solution:
    def productExceptSelf(self, nums: List[int]) -> List[int]:
        n = len(nums)
        leftProducts = [0] * n
        rightProducts = [0] * n
        product = 1
        for i in range(n):
            product *= nums[i]
            leftProducts[i] = product
        product = 1
        for i in range(n-1,-1,-1):
            product *= nums[i]
            rightProducts[i] = product
        result = [None] * n
        for i in range(n):
            result[i] = 1
            if i - 1 >= 0:
                result[i] *= leftProducts[i-1]
            if i + 1 < n:
                result[i] *= rightProducts[i+1]
        return result

class Solution:
    def productExceptSelf(self, nums: List[int]) -> List[int]:
        n = len(nums)
        result = [None] * n
        leftProduct = 1
        for i in range(n):
            result[i] = leftProduct
            leftProduct *= nums[i]
        rightProduct = 1
        for i in range(n-1,-1,-1):
            result[i] *= rightProduct
            rightProduct *= nums[i]
        return result

File: 0238-product-of-array-except-self/cpp/Solution.cpp
class Solution {
public:
    vector<int> productExceptSelf(vector<int>& nums) {
        int n = nums.size();
        vector<int> leftProduct(n);
        vector<int> rightProduct(n);
        int product = 1;
        for (int i = 0; i < n; ++i) {
            product *= nums[i];
            leftProduct[i] = product;
        }
        product = 1;
        for (int i = n - 1; i >= 0; --i) {
            product *= nums[i];
            rightProduct[i] = product;
        }

        vector<int> result(n);
        for (int i = 0; i < n; ++i) {
            result[i] = 1;
            if (i - 1 >= 0) {
                result[i] *= leftProduct[i - 1];
            }
            if (i + 1 < n) {
                result[i] *= rightProduct[i + 1];
            }
        }
        return result;
    }
};

class Solution {
public:
    vector<int> productExceptSelf(vector<int>& nums) {
        int n = nums.size();
        vector<int> result(n);
        int leftProduct = 1;
        for (int i = 0; i < n; ++i) {
            result[i] = leftProduct;
            leftProduct *= nums[i];
        }
        int rightProduct = 1;
        for (int i = n - 1; i >= 0; --i) {
            result[i] *= rightProduct;
            rightProduct *= nums[i];
        }
        return result;
    }
};
